### DISCUSSION TITLE
Serial communication between Cooja test file and ContikiMote

### DISCUSSION THREAD

**Discussion by fleutot** [contiki-ng/contiki-ng#1725] [Q&A]
I am trying to send data (or rather commands) from the Cooja test script, to the firmware in the simulated nodes. Note that I don't think this is the same as the border router example, I just want to implement the receiving end of the `write` in the Cooja javascript test file.

Node type: ContikiMote

Here is the mote type definition:
```
    <motetype>
      org.contikios.cooja.contikimote.ContikiMoteType
      <identifier>mtype345</identifier>
      <description>root node</description>
      <source>[CONFIG_DIR]/code/root.c</source>
      <commands>make root.cooja TARGET=cooja</commands>
      <moteinterface>org.contikios.cooja.interfaces.Position</moteinterface>
      <moteinterface>org.contikios.cooja.interfaces.Battery</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiVib</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiMoteID</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiRS232</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiBeeper</moteinterface>
      <moteinterface>org.contikios.cooja.interfaces.RimeAddress</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiIPAddress</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiRadio</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiButton</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiPIR</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiClock</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiLED</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiCFS</moteinterface>
      <moteinterface>org.contikios.cooja.contikimote.interfaces.ContikiEEPROM</moteinterface>
      <moteinterface>org.contikios.cooja.interfaces.Mote2MoteRelations</moteinterface>
      <moteinterface>org.contikios.cooja.interfaces.MoteAttributes</moteinterface>
      <symbols>false</symbols>
    </motetype>
```

I have implemented the serial-line.[ch] files into my project, they build fine. I call the init function, and I understand that I need to connect the actual input of characters into the serial-line module with a callback.

Here is what I am missing: what is this input interface? I've seen [someone suggesting using uart0](https://stackoverflow.com/a/50446837/108802), but I cannot find `uart0_set_input()` for ContikiMotes. I cannot find how to setup baudrate for that interface either.

I've tried with Cooja platform's rs232 (copied a bunch of files such as rs232.[ch], simEnvChange.[ch]), with no success (tried all the defined speeds):
```
    rs232_init();
    rs232_set_speed(RS232_19200);
    rs232_set_input(serial_line_input_byte);
    serial_line_init();
```
How do I connect and setup a ContikiMote serial input interface to the serial-input module, in the node's code?

**Comment by alexrayne** [comment]
cooja-platform uses rs232 as stdio - stdin/stdout.

cooja sim provide client/server tonel to IP port. just as PRL-board routed example shows.

you need some-how proxy your IP:port -> COMM device. i not experiensed in linux, maybe it trivial task there.

i not sure, but as i remember, saw some cooja-plugin or script for such toneling cooja -> comm

**Reply by alexrayne** [reply]
fast googling gaves [cooja-serial2pty](https://github.com/cmorty/cooja-serial2pty)

**Reply by fleutot** [reply]
Thanks, that means I was on the right path using rs232, I will continue with that.

I don't want to access the mote from other programs, which seems to be the goal of the project you linked. From the search results I've found, it seems you could write to motes directly from a Cooja test script with `write` without the need for a system device? Just like Cooja sees the motes' `printf` output.

**Reply by alexrayne** [reply]
not sure, yeas - as remember some test ans example scripts use it.

**Comment by atiselsts** [comment]
Hi, it's not super clear what you're done so far and what's not working. My first idea was to suggest serial2pty as well, but it's apparently what you need.

You should be able to call `node.write()`  from the Javascript judging by this code:
https://github.com/contiki-ng/cooja/blob/master/java/org/contikios/cooja/plugins/ScriptMote.java#L55

Did you try to do this and failed to receive anything in the Contiki C code?

**Reply by fleutot** [reply]
Yes, exactly!

**Reply by atiselsts** [reply]
do you ever get the `serial_line_input_byte` function called?

**Reply by fleutot** [reply]
No. I checked that the functions pointers are not NULL:
```    printf("init rs232\n");
    rs232_init();
    serial_line_init();
    //rs232_set_speed(RS232_19200);
    rs232_set_input(serial_line_input_byte);

    if (rs232_interface.doActionsBeforeTick == NULL) {
        printf("ERROR: rs232_interface Before\n");
    }
    if (rs232_interface.doActionsAfterTick == NULL) {
        printf("ERROR: rs232_interface After\n");
    }
```
I don't think I get to receiving anything on rs232 (I've tried to other speeds, but which one should it be?).

**Reply by alexrayne** [reply]
> (I've tried to other speeds, but which one should it be?)

on cooja target - speed not matter

**Reply by atiselsts** [reply]
You don't really need to call all these functions, the default initialization is already fine, and uses `serial_line_input_byte` as the input handler.

You need to debug the Cooja code, with some log messages or so.

**Reply by fleutot** [reply]
Thanks. I've narrowed it down a bit, I think. In `MoteInterfaceHandler.java`, `MoteInterfaceHandler::doActiveActionsBeforeTick()` does not seem to run for anything else than `ContikiClock`. 

I am not sure which one of Before or After is supposed to handle the communication Cooja -> mote, but since [rs232.c only implements the BeforeTick](https://github.com/contiki-ng/contiki-ng/blob/999b0d85f28c8f70b843542e216d6d62084e8563/arch/platform/cooja/dev/rs232.c#L110), I assumed this is where it is supposed to happen?

`MoteInterfaceHandler::doActiveActionsAfterTick()` _does_ run the ContikiRS232 interface (reasonable, since the class [implements the interface `PolledAfterActiveTicks`](https://github.com/contiki-ng/cooja/blob/815e9f980ac91fac3596910726c49cb931764067/java/org/contikios/cooja/contikimote/interfaces/ContikiRS232.java#L68)), but that's after tick. There nothing in the receiving end in rs232.c for after tick.

What confuses me now is: why does the `ContikiRS232` class implement the interface `PolledAfterActiveTicks`, while the receiving end `rs232` deals with data in `doInterfaceActionsBeforeTick`? Are these not the two sides of the communication Cooja -> ContikiMote? Is it expected that Cooja sends data after a tick, and the Mote consumes this data _before the **next** tick_?

**Reply by fleutot** [reply]
After a number of hours in java (which I don't know. Well, I know a lot more now), I finally reached the conclusions that:

- `SIM_INTERFACES(...)` was missing in my mote code. Adding `SIM_INTERFACES(&rs232_interface);` in my own mote code was required. I'm not sure this is supposed to be the case? `java/org/contikios/cooja/dialogs/CompileContiki.java` seems to try and do some magic here, that maybe should handle this automatically? Then why didn't do it in my case?
- EOL was not in the payload sent from Cooja to the mote with `write`. I modified `ContikiRS232:writeString()` to add 0x0a to the byte array `newData`, and I finally got the message through to my mote. I guess this is the wrong place to do it, but I don't know where the right place is. Who is to add EOL (0x0a)?

I expect that these things I did manually are normally done under the hood. But I don't know where, and how to make it happen?

**Reply by fleutot** [reply]
> EOL was not in the payload sent from Cooja to the mote with write

This was because I explicitly set the callback for the rs232 input. Calling this:
```
rs232_set_input(serial_line_input_byte);
```
sets up the rs232 to use the passed function as callback. In contrast, NOT setting the callback lets the rs232 module use `serial_line_input_byte` as default (so the function I want), but _also_ adds 0x0a (EOL) automatically.

So now instead of adding EOL in the java code, I just don't set the rs232 input callback and the message gets its EOL.

Incidentally, @atiselsts told me that I didn't need that init code, good call ;) (but I didn't think that having it would hurt... well it did.)

**Reply by fleutot** [reply]
Another thing I needed to do: explicitly call to `doActionsBeforeTick();` and `doActionsAfterTick();`. These functions are implemented in `simEnvChange.c`, which I build in my mote code.

Are these supposed to be called from `arch/platform/cooja/platform.c`?:
```
JNIEXPORT void JNICALL
Java_org_contikios_cooja_corecomm_CLASSNAME_tick(JNIEnv *env, jobject obj)
```

**Reply by atiselsts** [reply]
That's a quite lot of debugging you did here! Once/if you reach a working state, can you summarize the solution here so we can add it to the wiki?

**Comment by alexrayne** [comment]
>Are these supposed to be called from arch/platform/cooja/platform.c?:

exactly, anf them sure called on that platform.

if you need call them explicitly - signaling that you realy made some
strange, or broke contiky. cooja-target make by cooja sim on start, and it
works with no pribkems.


пт, 14 янв. 2022 г., 17:52 Gauthier Fleutot Östervall <
***@***.***>:

**Reply by fleutot** [reply]
So yes, I am working in a sandbox, which limited the implementation of Cooja motes. I'm sorry about that, I wasn't aware when I started this question.

As such, my case is probably not very relevant to other users.

For the sake of completeness, and as a summary:
- I did not link to platform.c
- To run anyway, I needed to add to my mote code:
```
extern const struct simInterface rs232_interface;
SIM_INTERFACES(&rs232_interface);
```
And a process to run the tick actions:
```
PROCESS_THREAD(process_interfaces_proc, ev, data)
{
    PROCESS_BEGIN();
    while (1) {
        static struct etimer timer;
        etimer_set(&timer, 10);
        PROCESS_WAIT_UNTIL(etimer_expired(&timer));
        doActionsBeforeTick();
        doActionsAfterTick();
    }
    PROCESS_END();
}
```
All this should not be needed if you build `platform.c`.

One thing that could be added to the wiki, maybe, is the fact that the serial-line callback adds EOL *if and only if* not setting the callback with `rs232_set_input`.

**Reply by atiselsts** [reply]
Ok, got it! Thanks for the summary and for the clarification that it's a special case because you did not have `platform.c` linked.
